### PR TITLE
[megatron] Fix redundant downloading of shards across workers/nodes for megatron dist checkpointing

### PR DESCRIPTION
**Pull request by erictang000**
## Summary

- **Per-rank cloud checkpoint loading for Megatron**: When loading a distributed checkpoint from S3/GCS, each rank now downloads only its own shard file(s) into a shared node-local temp directory, rather than every rank downloading the entire checkpoint. Local rank 0 creates the directory and downloads common metadata files; after a barrier, each rank downloads its shard in parallel. This reduces total download volume per node from `N * full_checkpoint` to `1 * full_checkpoint`.
- **New `io.download_file()` utility**: Adds a single-file cloud download helper (with S3 retry support) to complement the existing `download_directory()`.
- **New `test_save_load_checkpoint_cloud` test**: End-to-end test that saves a Megatron checkpoint to S3, loads it back via per-rank downloading, and verifies logit-exact reproducibility.

## Test plan

- [x] `test_save_load_checkpoint_cloud` -- saves checkpoint to S3, loads via per-rank shard download, verifies identical logits after reload + re-training
- [x] Existing `test_save_load_checkpoint` parametrizations (local filesystem) -- should be unaffected since the new cloud path is only triggered for `is_cloud_path()` URLs

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1414" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by erictang000** [comment]
Checkpointing load time for Qwen3-8B on 2 4xL40S nodes:

Before (~320s)
image

After (~200s)
image

File: tests/backends/skyrl_train/gpu/gpu_ci/test_save_load_checkpoint.py [hunk 1]
```diff
@@ -228,3 +232,81 @@ def test_save_load_checkpoint(ray_init_fixture, strategy, lora, fully_reshardabl
         if checkpoint_dir and os.path.exists(checkpoint_dir):
             print(f"Removing checkpoint directory: {checkpoint_dir}")
             shutil.rmtree(checkpoint_dir)
+
+
+@pytest.mark.megatron
+def test_save_load_checkpoint_cloud(ray_init_fixture):
```
Comment (SumanthRH) [inline tests/backends/skyrl_train/gpu/gpu_ci/test_save_load_checkpoint.py:238]: Hmm this test is meant specifically for our CI running on Anyscale. Can you skip this test if I'm running on a machine without `ANYSCALE_ARTIFACT_STORAGE` set?
